util/perl/TLSProxy/Proxy.pm

```diff
@@ -193,76 +191,119 @@ sub start
-            .($self->server_port)
-            ." -cert ".$self->cert." -cert2 ".$self->cert
-            ." -naccept ".$self->serverconnects;
-        unless ($self->supports_IPv6) {
-            $execcmd .= " -4";
-        }
-        if ($self->ciphers ne "") {
-            $execcmd .= " -cipher ".$self->ciphers;
-        }
-        if ($self->ciphersuitess ne "") {
-            $execcmd .= " -ciphersuites ".$self->ciphersuitess;
-        }
-        if ($self->serverflags ne "") {
-            $execcmd .= " ".$self->serverflags;
-        }
-        if ($self->debug) {
-            print STDERR "Server command: $execcmd\n";
-        }
-        exec($execcmd);
+    my $execcmd = $self->execute
+        ." s_server -max_protocol TLSv1.3 -no_comp -rev -engine ossltest"
+        ." -accept 0 -cert ".$self->cert." -cert2 ".$self->cert
+        ." -naccept ".$self->serverconnects;
+    unless ($self->supports_IPv6) {
+        $execcmd .= " -4";
+    }
+    if ($self->ciphers ne "") {
+        $execcmd .= " -cipher ".$self->ciphers;
+    }
+    if ($self->ciphersuitess ne "") {
+        $execcmd .= " -ciphersuites ".$self->ciphersuitess;
+    }
+    if ($self->serverflags ne "") {
+        $execcmd .= " ".$self->serverflags;
+    }
+    if ($self->debug) {
+        print STDERR "Server command: $execcmd\n";
+    }
+
+    open(my $savedin, "<&STDIN");
+
+    # Temporarily replace STDIN so that sink process can inherit it...
+    $pid = open(STDIN, "$execcmd |") or die "Failed to $execcmd: $!\n";
+
+    # Process the output from s_server until we find the ACCEPT line, which
+    # tells us what the accepting address and port are.
+    while (<>) {
+        print;
+        s/\R$//;                # Better chomp
+        next unless (/^ACCEPT\s.*:(\d+)$/);
+        $self->{server_port} = $1;
+        last;
     }
-    $self->serverpid($pid);
+
+    # Just make sure everything else is simply printed.
+    # The sub process simply inherits our STD* and will keep consuming
+    # server's output and printing it as long as there is anything there,
+    # out of our way.
+    if (eval { require Win32::Process; 1; }) {
+        Win32::Process::Create(my $h, "$^X", "perl -ne print", 0, 0, ".");
+        $self->serverpid($h->GetProcessID());
+    } else {
+        $pid = fork or exec("$^X -ne print");
+        $self->serverpid($pid);
+    }
+
+    # Change back to original stdin
+    open(STDIN, "<&", $savedin);
+    close($savedin);
+
+    print STDERR "Server responds on ",
+        $self->{server_addr}, ":", $self->{server_port}, "\n";
+
+    # Connect directly...
+    my $servaddr = $self->{server_addr};
+    $servaddr =~ s/[\[\]]//g; # Remove [ and ]
+
+    $self->{server_sock} = $IP_factory->(PeerAddr => $servaddr,
+                                         PeerPort => $self->{server_port},
+                                         Proto => 'tcp')
+                           or die "unable to connect: $!\n";
```

Comment:
Also, wouldn't it make more sense to have this early in `clientstart`...

Reply:
It would, but I use connection as synchronization point at the end of clientstart (which should really be called clientrun or something similar). One of the things in this request is that all sleeps are omitted. They were meant to work around race conditions, but they don't actually resolve them. Synchronization does. Well, one can still argue that there is race condition at the point when connect_to_server is called at the end of clientstart. To remind, it was race between outputs with TAP(?) output blending into output from s_server and causing parsing and no-plan errors. Race might still be possible, because output from s_server is passed to `$^X -ne print` and we have no way to synchronize on that. Though it gets explicitly broken to lines, which might be sufficient... I've updated commentary and remove code duplication...

Reply:
In other words it *is* kind of controversial and one can say that it, connection to s_server, should be moved back into beginning of clientstart. As for s_server's output being explicitly broken to lines [by `$^X -ne print`], i.e. each line is performed by separate write(2). Related question also is if TAP output is broken to lines, in same meaning. If it isn't, then I'd say that it should be... Where does one look?

Reply:
TAP output from our programs?  That's in `test/testutil` for test programs...  for `apps/openssl`, that should already be there for bio_out, but it just dawned on me that `s_client` and `s_server` use a separate BIO for their output...  probably for good reasons too.  However, you also have TLSProxy itself that outputs quite a lot in verbose mode, and that output isn't adapted to TAP either.

Reply:
> That's in test/testutil for test programs...

In this case TAP output is not generated by any of the test/program. I would guess it's generated by perl. And that's what I'm wondering about. As for test/program, it's not an issue, because there is only one program generating output, while *here* we have multiple programs doing that, s_server, s_client and script itself. For *this* to work [TAP-wise] one either have to synchronize execution (done for s_client and **last** connection to s_server), or ensure that lines are not blended with each other. At least that's what I figure is sufficient to keep TAP parser happy. If two programs issue write(2) calls against pipe they should come out on the other side as separate chunks. Trouble comes when one program collects output in buffer and then issues write(2) with multiple lines and potentially partial line. And that by the way what was happening with output from s_server [that lead to sleep introduction]. Hence the importance of not issuing partial lines. Now output from s_server is broken to lines [by $^X -ne print], and remaining question is about TAP output from script, the part that generates it. (Just in case, output from s_client is not an issue, because it finishes synchronously, s_server is the only one running in background without option to synchronize on its output.)